include/v4l2_camera/rate_bound_status.hpp

```diff
@@ -0,0 +1,281 @@
+                    "Rate is reasonable") {}
+  };
+
+  struct Warn : public StateBase
+  {
+    Warn()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::WARN,
+                    "Rate is within warning range") {}
+  };
+
+  struct Error : public StateBase
+  {
+    Error()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::ERROR,
+                    "Rate is out of valid range") {}
+  };
+
+  using StateHolder = std::variant<Stale, Ok, Warn, Error>;
+
+ public:
+  /**
+   * \brief Constructs RateBoundstatus, which inherits diagnostic_updater::DiagnosticTask.
+   *
+   * \param ok_params The pair of min/max frequency for the topic rate to be recognized as "OK".
+   * \param warn_params The pair of min/max frequency for the topic rate to be recognized as "WARN".
+   * These values should have a wider range than `ok_params`.
+   * \param num_frame_transition The number of the successive observations for the status
+   * transition. E.g., the status will not be changed from OK to WARN until successive
+   * `num_frame_transition` WARNs are observed.
+   * \param immediate_error_report If true (default), errors related to the rate bounds will be
+   * reported immediately once observed; otherwise, the hysteresis damping method using
+   * `num_frame_transition` will be adopted
+   * \param name The arbitrary string to be assigned for this diagnostic task.
+   * This name will not be exposed in the actual published topics.
+   */
+  RateBoundStatus(const RateBoundStatusParam& ok_params,
+                  const RateBoundStatusParam& warn_params,
+                  const size_t num_frame_transition = 1,
+                  const bool immediate_error_report = true,
+                  std::string name = "rate bound check")
+      : DiagnosticTask(name), ok_params_(ok_params), warn_params_(warn_params),
+        num_frame_transition_(num_frame_transition),
+        immediate_error_report_(immediate_error_report), zero_seen_(false),
+        candidate_state_(Stale{}), current_state_(Stale{})
+  {
+    if (num_frame_transition < 1) {
+      num_frame_transition_ = 1;
+    }
+
+    // Confirm `warn_params` surely has wider range than `ok_params`
+    if (!(warn_params_.min_frequency < ok_params_.min_frequency &&
+          ok_params_.max_frequency< warn_params_.max_frequency)) {
+      throw std::runtime_error(
+          "Invalid range parameters were detected. warn_params should specify a range "
+          "that includes a range of ok_params.");
+    }
+  }
+
+  /**
+   * \brief Calculate the frequency of how much this function is called
+   *
+   */
+  void tick()
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    double stamp = std::chrono::duration<double>(
+        std::chrono::steady_clock::now().time_since_epoch()).count();
+
+    if (!previous_frame_timestamp_) {
+      previous_frame_timestamp_ = stamp;
+    }
+
+    double delta = stamp - previous_frame_timestamp_.value();
+    if (delta == 0) {
```

Comment:
It seems like the delta == 0 case only happens when `previous_frame_timestamp_` has not been initialized. So would this be easier to understand?
```
if (!previous_frame_timestamp_) {
  zero_seen_ = true;
} else {
  zero_seen_ = false;
  double delta = stamp - previous_frame_timestamp_.value();
  frequency_ = (delta < 10 * std::numeric_limits<double>::epsilon()) ?
                   std::numeric_limits<double>::infinity() : 1. / delta;
}
previous_frame_timestamp_ = stamp;
```

Reply:
Thank you for deep checking. I simplified here in f5ab76c as you suggested!